### PR TITLE
[DRAFT] Adds support for mixed precision NVIDIA A100 Tensor Cores (F32 <= BF16 * BF16 + F32)

### PR DESCRIPTION
**Pull request by manishucsd**
This PR adds support for mixed precision NVIDIA A100 Tensor Core support addressing the issue https://github.com/openxla/iree/issues/13414 from Epic #13812

BF16 input for lhs, and BF16 input for rhs, F32 for accumulation, and F32 output data types can be supported on CUDA backend by lowering to `mma.sync.aligned.m16n8k16.row.col.f32.bf16.bf16.f32` ptx instruction. Inputs and outputs are different bit-width (16b inputs and 32b outputs)

Prerequisites for merging this PR:
1. Merge #13857
2. Rebase
3. Resolve the below error

Fails the at createIntrinsicCall on nvvm.mma.sync on `bf16`

```bash
#20 0x00007fa1561cbe62 mlir::LLVM::detail::createIntrinsicCall(llvm::IRBuilderBase&, unsigned int, llvm::ArrayRef<llvm::Value*>, llvm::ArrayRef<llvm::Type*>) /mnt/disks/gcloud_workspace/repos/iree/iree_tree_2/iree/third_party/llvm-project/mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:586:3

```

```mlir
%489 = nvvm.mma.sync A[%293, %295, %297, %299]  B[%392, %394]  C[%484, %485, %487, %488]  {layoutA = #nvvm.mma_layout<row>, layoutB = #nvvm.mma_layout<col>, multiplicandAPtxType = #nvvm.mma_type<bf16>, multiplicandBPtxType = #nvvm.mma_type<bf16>, shape = #nvvm.shape<m = 16, n = 8, k = 16>} : (vector<2xbf16>, vector<2xbf16>, f32) -> !llvm.struct<(f32, f32, f32, f32)>
```

### CODE REVIEW

**Comment by manishucsd** [comment]
Currently, I am seeing a failure when compiling BF16 input and F32 output during createIntrinsicCall on 
```mlir
nvvm.mma.sync A[%293, %295, %297, %299]  B[%392, %394]  C[%484, %485, %487, %488]  {layoutA = #nvvm.mma_layout<row>, layoutB = #nvvm.mma_layout<col>, multiplicandAPtxType = #nvvm.mma_type<bf16>, multiplicandBPtxType = #nvvm.mma_type<bf16>, shape = #nvvm.shape<m = 16, n = 8, k = 16>} : (vector<2xbf16>, vector<2xbf16>, f32) -> !llvm.struct<(f32, f32, f32, f32)>
```

F16 input and F32 output goes and the nvvm at the same stage in the compilation pipeline looks like as follows:

```mlir
nvvm.mma.sync A[%1231, %1232, %1233, %1234]  B[%1235, %1236]  C[%1239, %1241, %1244, %1246]  {layoutA = #nvvm.mma_layout<row>, layoutB = #nvvm.mma_layout<col>, shape = #nvvm.shape<m = 16, n = 8, k = 16>} : (vector<2xf16>, vector<2xf16>, f32) -> !llvm.struct<(f32, f32, f32, f32)>
```

Please find full dumps for these cases [here](https://drive.google.com/drive/folders/1e7agAR4GZZqmcM-KXFybfLeWiVEDCcJX?usp=share_link)

dump_f32_f16.mlir (F16 input, F32 accumulation. This compiles and works file)
dump_f32_bf16.mlir (BF16 input, F32 accumulation. Fails at createIntrinsicCall)

cc: @MaheshRavishankar , @rsuderman Does anything ring a bell. You can check the IREE and LLVM upstream changes in this PR. I have made some NVVMDialect changes to get to this point, maybe some more changes are needed.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 1]
```diff
@@ -518,8 +518,8 @@ static void addLowerToLLVMGPUPasses(OpPassManager &pm, bool useROCM) {
   pm.addPass(createConvertComplexToStandardPass());
 
   // Convert BF16 operations to occur as F32.
-  pm.addPass(createConvertBf16ArithToF32Pass());
-  pm.addPass(createConvertBf16ToUInt16BuffersPass());
+  // pm.addPass(IREE::Util::createPromoteArithBF16ToF32Pass());
```
Comment (mattwalsh) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:521]: Do you mean to leave this in as commented?  Or simply eliminate?  if you want to keep it, please add a comment as to what this commented out code might help the reader understand / how it might be used later.

File: experimental/dispatch_profiler/manifest.py [hunk 2]
```diff
@@ -200,8 +200,8 @@ def append(self, dispatch_collection_list):
   def initialize(self):
     """Initialize the mainfest object by generating dispatches for supported operations."""
     self.append(CudaMatmulGenerator(self.args).generate())
-    self.append(CudaSplitKMatmulGenerator(self.args).generate())
-    self.append(CudaBatchMatmulGenerator(self.args).generate())
+    #self.append(CudaSplitKMatmulGenerator(self.args).generate())
```
Comment (mattwalsh) [inline experimental/dispatch_profiler/manifest.py:203]: Similar as above, elide or document with a comment what this might be used for in future

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 1 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:521]: Yes, ultimately this should go away. This is still a DRAFT PR. Leaving the comment unresolved to handle it before merging.

File: experimental/dispatch_profiler/manifest.py (hunk 2 above)
Comment (manishucsd) [reply experimental/dispatch_profiler/manifest.py:203]: Sam as above. Just focusing on BF16 matmul before the PR is ready to merge.

**Comment by manishucsd** [comment]
BF16 inputs, F32 accumulation, and F32 output works functionally, performance is terrible. 

```
python3 dispatch_profiler/profiler.py --dispatches=matmul_3456x2048x1024_bf16t_bf16t_f32t
---------------------------------------------------------------- 
Dispatch      : matmul_3456x2048x1024_bf16t_bf16t_f32t_tile_config_128x256_32x3_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x2048x1024_bf16t_bf16t_f32t
Configuration : tile_config_128x256_32x3_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=2048 --k=1024 --lhs=bf16t --rhs=bf16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.259
GFLOPs        : 55967.24
---------------------------------------------------------------- 
```

Concerns and more work TODO:
1) The performance is terrible with BF16 it should be same as with [F16 input, F32 accum, F32 output](https://github.com/openxla/iree/pull/13857#issue-1733374505), i.e., 183 TFLOPs and not 55 TFLOPs. Ideally it should be over 200 TFLOPs. Need to run cuBLAS and CUTLAS get the limits of hand-written kernels. 
2) Also, the path/lowering has too many steps from NVGPU to PTX. Also, till NVGPU and we need to rethink it these lowerings: matul description -> nvgpu -> ptx. We need to rethink the lowering from matmul desc to ptx.

**Comment by manishucsd** [comment]
Addressing 1) from the [comment](https://github.com/openxla/iree/pull/13858#issuecomment-1581246646) with [e825d76](https://github.com/openxla/iree/pull/13858/commits/e825d76c6b32b965a408c5b5cac603fa337b3f1c). This change of few characters held back 158 TFLOPs, Thus adding 158 TFLOPs to initial functional BF16 run of 55 TFLOPs, resulting in 213 TFLOPs. 

We were missing `cp.async` and latency-tolerant fine-grained instruction scheduling was getting missed on BF16 data type. 

```bash
 $ python3 dispatch_profiler/profiler.py --dispatches=matmul_3456x1024x2048_bf16t_bf16t_f32t_tile_config_128x256_32x3_tensorcore_mmasync
---------------------------------------------------------------- 
Dispatch      : matmul_3456x1024x2048_bf16t_bf16t_f32t_tile_config_128x256_32x3_tensorcore_mmasync
Provider      : IREE Codegen
OpKind        : OperationKind.Matmul
Operation     : matmul_3456x1024x2048_bf16t_bf16t_f32t
Configuration : tile_config_128x256_32x3_tensorcore_mmasync
Arguments     : --batch_count=1 --m=3456 --n=1024 --k=2048 --lhs=bf16t --rhs=bf16t --result=f32t
                --split_k_mode=N/A --split_k_slices=N/A
Verification  : SUCCESS
Runtime(ms)   : 0.068
GFLOPs        : 213169.33

```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 1 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:521]: The alternative would be to add a command line option around this.
That could actually be useful for debugging purposes.
Up to you.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUUtils.cpp [hunk 3]
```diff
@@ -191,7 +191,8 @@ void createAsyncGroups(RewriterBase& rewriter, func::FuncOp funcOp,
 
     VectorType vecType = llvm::cast<VectorType>(vectorVal.getType());
     if (!((vecType.getElementType().isF32() && vecType.getNumElements() <= 4) ||
-          (vecType.getElementType().isF16() &&
+          ((vecType.getElementType().isF16() ||
+            vecType.getElementType().isBF16()) &&
            vecType.getNumElements() <= 8))) {
       LDBG("----readOp is not (<=4)xf32 or (<=8)xf16 -> Skip");
```
Comment (qcolombet) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUUtils.cpp:197]: Nit: update the string in the debug output: `f16` => `(b)f16`.

File: experimental/dispatch_profiler/library.py [hunk 4]
```diff
@@ -6,6 +6,7 @@
 
 import enum, re
 from enum import auto
+from ml_dtypes import bfloat16
```
Comment (qcolombet) [inline experimental/dispatch_profiler/library.py:9]: The dispatch_profiler changes could be there own PR, right?
IIUC, IREE should still support BF16 but it will promote them to FP32. Which is fine for landing the dispatch profiler changes.

**Review by qcolombet — COMMENTED** [review]
When the LLVM support for the intrinsic lands, make sure to add a test case to make sure we generate the right instruction.

If the intrinsic part takes too much time, can we go with inline asm in the meantime?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 1 above)
Comment (qcolombet) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:521]: Following yesterday's meeting, I think we should put these passes behind a flag.
Have this flag currently turned on (i.e., the passes run by default).
And flip the default for that flag when the LLVM changes land.
That way we can get this code in and we will see the performance boost in the benchmarks when we flip the flag.

**Comment by qcolombet** [comment]
@manishucsd what do I need to apply locally to be able to use `BF16`?

I applied this PR and ran something simpler (`BF16 <= BF16 * BF16 + BF16`) and it crashes for me:
```
<unknown>:0: error: MMAMatrixType elements must be SI8, UI8, I32, F16, or F32 
iree-compile: third_party/llvm-project/mlir/include/mlir/IR/StorageUniquerSupport.h:174: static ConcreteT mlir::detail::StorageUserBase<mlir::gpu::MMAMatrixType, mlir::Type, mlir::gpu::MMAMatrixStorageType, mlir::detail::TypeUniquer>::get(mlir::MLIRContext *, Args...) [ConcreteT = mlir::gpu::MMAMatrixType, BaseT = mlir::Type, StorageT = mlir::gpu::MMAMatrixStorageType, UniquerT = mlir::detail::TypeUniquer, Traits = <>, Args = <llvm::ArrayRef<long>, mlir::Type, llvm::StringRef>]: Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.    
```
